### ISSUE TITLE
[Service Bus] Thread safety on sender/receiver

### ISSUE THREAD

**Issue by yunhaoling** [Azure/azure-sdk-for-python#21385]
context: service bus sender/receiver is not thread safe, which will result in race condition when run in parallel execution, and the internal status would be corrupted

question: do we want our sender/receiver to be thread safe?

- if yes, then we should fix it
  - concern for thread safety
      - performance
      - sdk implementation guideline? e.g. in general, we don't guarantee thread-safety
- if no, then we'd better document it

----------

to sum up, in general our SDK should be thread safe, which means we need thread safety on our service bus sender and receiver.

### scopes

- thread safe sync sender and receiver
- concurrency safe async sender and receiver

**Comment by tsabsch** [comment]
Hi, we are using `azure-servicebus` as part of our FastAPI+uvicorn application, and are affected by this. A short documentation about thread-safety would be very appreciated, especially since [1] suggests that the SDK actually _is_ thread-safe:

> The following note applies to all SDKs:
> Establishing a connection is an expensive operation that you can avoid by reusing the same factory or client objects for multiple operations. You can safely use these client objects for concurrent asynchronous operations and from multiple threads.

[1] https://docs.microsoft.com/en-us/azure/service-bus-messaging/service-bus-performance-improvements?tabs=net-standard-sdk-2#reusing-factories-and-clients
(I'm aware by now that the referenced page seems to only apply to .NET, but it isn't obvious if you're on it)

**Comment by QianYC** [comment]
Hi I wonder do we have a conclusion about this thread? Do users need to take care of race conditions when using sender/receiver concurrently?

**Comment by swathipil** [comment]
Hi @QianYC - This library is not thread-safe currently, so you will need to consider race conditions. Thanks.
